### PR TITLE
fix(agent): preserve drafts and stabilize tool follow-ups

### LINKED ISSUE #246: fix(agent): prevent failed re-extraction from corrupting draft state after analysis failure
## 问题描述

分析失败后，Agent 会反复调用 `extract_draft_params` 尝试修复，但当重新识别失败时，会把已有有效 draft 覆盖成：

- `inferredType: "unknown"`
- `skillId: "generic"`
- `criticalMissing: ["inferredType"]`

之后 `build_model` 甚至会返回 0 节点、0 单元的空模型，并且仍显示 `success: true`。

## 复现现象

1. 简支梁前一次建模/分析可用。
2. 用户继续要求改跨度或重新执行静力分析。
3. `run_analysis` 失败。
4. Agent 调用 `extract_draft_params`。
5. draft 被覆盖为 `unknown/generic`。
6. 再调用 `build_model` 得到：

```json
{
  "success": true,
  "nodeCount": 0,
  "elementCount": 0
}
```

## 当前问题

- `extract_draft_params` 在识别失败或退回 generic/unknown 时，会污染已有有效 draft state。
- 分析失败后的恢复策略不稳定：Agent 会反复重新提取，而不是基于已有 draft/model 做局部修正或明确说明无法自动修复。
- `build_model` 对 `unknown/generic` 的空状态仍返回 `success: true`，导致用户误以为空模型是有效结果。
- 对话层会把问题解释成“缺少 inferredType”或“草稿损坏”，但没有阻止状态继续恶化。

## 期望行为

- 如果已有 `draftState` 是有效结构类型（如 beam），新的 `extract_draft_params` 识别失败时，不应覆盖为 `unknown/generic`。
- 对于低置信度或 fallback 识别结果，应保留上一版有效 draft，并返回可解释的 warning / clarification。
- `build_model` 不应对空模型返回 `success: true`。
- 分析失败后的自动恢复应优先做受控修复，例如：
  - 保留结构类型与已有参数；
  - 只更新用户明确修改的字段；
  - 如果需要重新建模，先确认或报告具体缺失项；
  - 不要反复全量覆盖 draft。

## 建议修复

- 在 `extract_draft_params` 中加入 state preservation guard：
  - 当 `existingState.inferredType` 有效，且新识别结果为 `unknown/generic` 时，默认不覆盖结构类型。
  - 仅合并明确提取出的参数字段。
- 给 `build_model` 增加空模型失败保护：
  - `nodes.length === 0` 或 `elements.length === 0` 时返回失败，而不是 `success: true`。
- 增加分析失败恢复策略测试：
  - 有效 beam draft + 后续模糊消息/失败重试，不应退化为 unknown。
  - unknown/generic fallback 不应删除已有有效 beam 参数。
  - 空模型不能作为成功结果返回。

## 验收标准

- 复现对话中，分析失败后的再次提取不会把 beam draft 覆盖成 unknown/generic。
- `build_model` 不再返回 0 节点 0 单元的成功模型。
- 用户继续追问“缺啥东西”时，系统能说明真实失败原因或具体缺失项，而不是把状态污染后的 `inferredType` 当作唯一问题。
- 增加后端回归测试覆盖上述场景。

### PR DESCRIPTION
**Pull request by qinsz01**
## Summary

Describe the problem and fix in 2-5 bullets:

- Problem: failed or low-confidence re-extraction could overwrite a previously valid draft with `unknown/generic`, and follow-up tool turns could silently end after tools because streaming message chunks were stored as final graph messages.
- Why it matters: users could see a valid structural model degrade into a 0-node/0-element model, lose explicit follow-up tool input, or get no final explanation after tool execution.
- What changed: preserve valid draft state on unknown fallback, reject empty generic models, prefer explicit tool messages over raw last-user follow-ups, normalize/recover empty tool-final replies, and disable internal LangChain streaming for the agent graph state so final messages are complete `AIMessage` instances.
- What did NOT change (scope boundary): no frontend behavior, schema, persistence, permissions, or analysis engine logic was changed.

## Change Type (select all)

- [x] Bug fix
- [ ] Feature
- [ ] Refactor required for the fix
- [ ] Docs
- [ ] Security hardening
- [ ] Chore/infra

## Scope (select all touched areas)

- [x] Gateway / orchestration
- [x] Skills / tool execution
- [ ] Auth / tokens
- [ ] Memory / storage
- [ ] Integrations
- [x] API / contracts
- [ ] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR

- Closes #246
- Related #
- [x] This PR fixes a bug or regression

## Root Cause / Regression History (if applicable)

- Root cause: the agent recovery path treated fallback `unknown/generic` extraction as authoritative and allowed it to replace valid draft state; additionally, graph `messages` stream mode made LangChain prefer streaming inside `model.invoke()`, so GLM-compatible streaming chunks could be written into graph state as blank/generic final messages.
- Missing detection / guardrail: no guard prevented low-confidence extraction from deleting valid draft fields; generic `build_model` considered empty models successful; explicit tool input could be overwritten by `lastUserMessage`; graph state did not force complete non-streaming model messages.
- Prior context (`git blame`, prior PR, issue, or refactor if known): reported in #246 from repeated `extract_draft_params` / `build_model` recovery attempts after analysis failure.
- Why this regressed now: issue #246 exposed the recovery path with follow-up turns such as retry/clarification messages, where the latest user utterance was not a complete structural request.
- If unknown, what was ruled out: ruled out frontend rendering as the blank-message source; persisted checkpoints already contained empty `ChatMessageChunk` final messages.

## Regression Test Plan (if applicable)

- Coverage level that should have caught this:
  - [x] Unit test
  - [x] Seam / integration test
  - [ ] End-to-end test
  - [ ] Existing coverage already sufficient
- Target test or file: `backend/src/agent-langgraph/__tests__/draft-preservation.test.mjs`, `backend/src/agent-langgraph/__tests__/graph-empty-response.test.mjs`, `backend/src/utils/__tests__/llm-options.test.mjs`
- Scenario the test should lock in: valid draft state must survive `unknown/generic` fallback; generic empty models must fail; explicit tool messages must be used; blank final responses after tools must be guarded; graph LLM construction can disable internal streaming.
- Why this is the smallest reliable guardrail: these tests cover the state transitions and model options directly without requiring a live LLM provider.
- Existing test that already covers this (if any): existing stream contract coverage remains relevant and was rerun.
- If no new test is added, why not: N/A

## User-visible / Behavior Changes

- Follow-up recovery should keep the last valid structural draft instead of degrading to `unknown/generic`.
- Empty generic models now fail instead of reporting `success: true` with 0 nodes / 0 elements.
- Tool follow-ups that carry explicit synthetic context should no longer be replaced by vague user messages such as “你说？”.
- After tools run, the conversation should not silently end with a blank final assistant message.

## Diagram (if applicable)

```text
Before:
[user follow-up] -> [fallback extraction unknown/generic] -> [valid draft overwritten]
[tools complete] -> [streaming ChatMessageChunk(content="")] -> [silent final response]

After:
[user follow-up] -> [fallback extraction warning] -> [valid draft preserved]
[tools complete] -> [complete non-streaming AIMessage] -> [visible final response]
```

## Security Impact (required)

- New permissions/capabilities? (`Yes/No`) No
- Secrets/tokens handling changed? (`Yes/No`) No
- New/changed network calls? (`Yes/No`) No
- Command/tool execution surface changed? (`Yes/No`) No
- Data access scope changed? (`Yes/No`) No
- If any `Yes`, explain risk + mitigation: N/A

### Environment

- OS: Windows / PowerShell local dev workspace
- Runtime/container: Node backend test runner
- Model/provider: GLM-compatible OpenAI endpoint was the reported runtime; tests use no live provider
- Integration/channel (if any): LangGraph chat stream
- Relevant config (redacted): API keys redacted; default backend test configuration

### Steps

1. Start from a valid structural draft and trigger a failed/low-confidence re-extraction with a vague follow-up.
2. Call `build_model` after the fallback extraction.
3. Execute a tool-follow-up turn where the tool input carries explicit context while `lastUserMessage` is vague.
4. Complete a tool turn where the final LLM response may be empty.

### Expected

- Existing valid draft fields remain available.
- Empty generic models are not reported as successful.
- Explicit tool input is preserved.
- Final assistant response is visible and graph state stores complete assistant messages.

### Actual

- Covered by regression tests added in this PR.

## Evidence

Attach at least one:

- [x] Failing test/log before + passing after
- [x] Trace/log snippets
- [ ] Screenshot/recording
- [ ] Perf numbers (if relevant)

Verification commands run successfully:

```text
npm test --prefix backend -- --runInBand src/utils/__tests__/llm-options.test.mjs src/agent-langgraph/__tests__/graph-empty-response.test.mjs src/agent-langgraph/__tests__/draft-preservation.test.mjs src/agent-langgraph/__tests__/streaming.test.mjs
npm run lint --prefix backend
node tests/runner.mjs validate validate-chat-stream-contract
npm test --prefix backend -- --runInBand
```

## Human Verification (required)

- Verified scenarios: target backend regressions, backend lint, chat stream contract, full backend Jest suite.
- Edge cases checked: low-confidence fallback over valid state, empty generic build result, explicit tool input versus vague last user message, blank final response after tool calls, graph LLM `disableStreaming` option.
- What you did **not** verify: live browser UI replay and live production LLM provider replay after the fix.

## Review Conversations

- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

If a bot review conversation is addressed by this PR, resolve that conversation yourself. Do not leave bot review conversation cleanup for maintainers.

## Compatibility / Migration

- Backward compatible? (`Yes/No`) Yes
- Config/env changes? (`Yes/No`) No
- Migration needed? (`Yes/No`) No
- If yes, exact upgrade steps: N/A

## Risks and Mitigations

- Risk: disabling internal streaming for the agent graph may reduce token-by-token final-answer streaming from that node.
  - Mitigation: chat stream contract still passes; tool/update events continue flowing, and graph state correctness is prioritized. If needed, token display streaming can be reintroduced separately without writing chunks into persistent graph state.

### CODE REVIEW

File: backend/src/agent-langgraph/graph.ts [hunk 1]
```diff
@@ -58,7 +117,11 @@ function createCallModelNode(
   ): Promise<Partial<AgentState>> {
     const log = getAgentLogger(config);
     const configurable = config.configurable as Partial<AgentConfigurable> | undefined;
-    const model = createChatModel(0);
+    // LangGraph's `messages` stream mode installs a callback handler that
+    // prefers streaming. LangChain may then make `invoke()` return message
+    // chunks; GLM-compatible streaming can yield empty/generic chunks after
+    // tool calls. Keep graph state on complete AIMessage instances.
+    const model = createChatModel(0, { disableStreaming: true });
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-langgraph/graph.ts:124]: Addressed in `7de1a3a`: the graph call still uses `disableStreaming` because the installed `@langchain/openai` supports it and LangChain's callback-preferred streaming path checks that flag, but the options now also pass `streaming: false` when graph-state streaming is disabled so the constructor intent is explicit and aligned with the standard streaming field.

File: backend/src/utils/llm.ts [hunk 2]
```diff
@@ -9,26 +9,38 @@ type ChatModelConfigLike = Pick<
...
+  temperature: number,
+  runtimeOptions: ChatModelRuntimeOptions = {},
+) {
   return {
     modelName: modelConfig.llmModel,
     temperature,
     timeout: modelConfig.llmTimeoutMs,
     maxRetries: modelConfig.llmMaxRetries,
     apiKey: modelConfig.llmApiKey,
+    disableStreaming: runtimeOptions.disableStreaming ?? false,
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/utils/llm.ts]: Addressed in `7de1a3a`: `buildChatModelOptions()` now sets both `disableStreaming: true` and `streaming: false` when the runtime option requests non-streaming graph-state responses. The default path keeps `streaming` unset so normal callers keep their existing behavior.

File: backend/src/agent-langgraph/tools.ts [hunk 3]
```diff
@@ -77,6 +86,72 @@ function buildDraftProgress(
...
+  plugin?: AgentSkillPlugin | null;
+  locale: 'zh' | 'en';
+}): {
+  responseJson: Record<string, unknown>;
+  stateUpdate: Partial<AgentState>;
+} {
+  const plugin = args.plugin ?? null;
+  const nextState = plugin
+    ? plugin.handler.mergeState(args.existingState, {})
+    : { ...args.existingState, updatedAt: Date.now() };
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-langgraph/tools.ts]: Addressed in `7de1a3a`: preserved draft extraction now refreshes `updatedAt` after the plugin/no-plugin merge path in both branches, and the regression test asserts the preserved state timestamp advances.

File: backend/src/agent-langgraph/tools.ts [hunk 4]
```diff
@@ -77,6 +86,72 @@ function buildDraftProgress(
...
+  const missing = plugin
+    ? plugin.handler.computeMissing(nextState, 'execution')
+    : { critical: [], optional: [] };
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-langgraph/tools.ts]: Addressed in `7de1a3a`: the no-plugin preservation path is now conservative. It keeps the valid draft but reports `criticalMissing: ['skillPlugin']`, returns `canProceed: false`, and asks for clarification instead of allowing `build_model` when missing fields cannot be computed reliably. The draft preservation test was updated for this case.

**Review by guyi2000 — APPROVED** [review]
